enterprise_gateway/services/kernels/remotemanager.py

```diff
@@ -186,6 +190,50 @@ async def start_kernel(self, *args, **kwargs):
         self.parent.kernel_session_manager.create_session(kernel_id, **kwargs)
         return kernel_id
 
+    async def restart_kernel(self, kernel_id):
+        kernel = self.get_kernel(kernel_id)
+        self.log.debug(f"Current value of the Kernel Restarting: {kernel.restarting}")
+        if kernel.restarting:  # assuming duplicate request.
+            await self.wait_and_poll_for_restart_to_complete(kernel_id, "restart")
+            self.log.info(
+                f"Done with waiting for restart to complete. Current value of restarting: {kernel.restarting}. "
+                f"Skipping restarting kernel.")
+            return
+        self.log.info("Going ahead to process restart kernel request.")
+        try:
+            kernel.restarting = True  # Moved in out of RemoteKernelManager
+            await super(RemoteMappingKernelManager, self).restart_kernel(kernel_id)
+        finally:
+            self.log.debug("Resetting `restarting` flag to False.")
+            kernel.restarting = False
+
+    async def shutdown_kernel(self, kernel_id, now=False, restart=False):
+        kernel = self.get_kernel(kernel_id)
+        self.log.debug(f"Current value of the Kernel Restarting: {kernel.restarting}")
+        if kernel.restarting:
+            await self.wait_and_poll_for_restart_to_complete(kernel_id, "shutdown")
+        try:
+            await super(RemoteMappingKernelManager, self).shutdown_kernel(kernel_id, now, restart)
+        except KeyError as ke:  # this is hit for multiple shutdown request.
+            self.log.exception(f"Exception while shutting Kernel: {kernel_id}: {ke}")
+
+    async def wait_and_poll_for_restart_to_complete(self, kernel_id, action="shutdown"):
+        kernel = self.get_kernel(kernel_id)
+        start_time = int(time.time())  # epoc time in seconds
```

Comment:
We should tolerate milliseconds since we'll want to allow for a sub-second interval.
```suggestion
        start_time: float = time.time()  # epoc time
```